### DISCUSSION TITLE
Handle errors in getStaticProps/getStaticPaths

### DISCUSSION THREAD

**Discussion by balazsorban44** [vercel/next.js#11969] [Help] [answered]
I discussed in #11180, I wish to log eventual data fetching errors to an external service in  getStaticProps/getStaticPaths
Here is my code:

```ts


type GSPContext = "getStaticProps" | "getStaticPaths"
/** Creates and returns an ApolloClient for builds */
const getClient = async (context: GSPContext) => {
  log.info({ context }, "{context}: creating ApolloClient")
  const token = await getServiceAccountAccessToken()
  const client = new ApolloClient({
    credentials: "same-origin",
    uri: process.env.APP_CONTENT_API,
    cache: new InMemoryCache({ fragmentMatcher }),
    request: (operation) => {
      operation.setContext({
        headers: {
          authorization: `Bearer ${token}`
        }
      })
    }
  })
  log.info({ context }, "{context}: ApolloClient created")
  return client
}

export const getStaticPaths: GetStaticPaths = async () => {
  const variables = {
    url: "/articles"
  }

  log.info(variables, "getStaticPaths: invoking at {url}")
  try {
    const client = await getClient("getStaticPaths")
    log.info(variables, "getStaticPaths: querying ApolloClient at {url}")

    const { data, errors } = await client.query<PagePathsQuery>({
      query: PagePathsDocument, variables
    })

    if (errors) {
      log.error(errors, "getStaticPaths: GraphQLErrors occured")
      throw errors
    }

    const { pages } = data

    if (!pages) {
      throw { message: "getStaticPaths: Pages not found at {url}" }
    }

    log.info(variables, "getStaticPaths: finished at {url}")
    return ({
      paths: pages.map(page => ({ params: { slug: [page.url] } })),
      fallback: true
    })
  } catch (error) {
    const message = error.message || "getStaticPaths: failed at {url}"
    log.error({ ...error, ...variables }, message)
    throw error
  }
}

export const getStaticProps: GetStaticProps = async (ctx) => {
  const { slug } = (ctx.params as ParsedUrlQuery)
  const url = Array.isArray(slug) ? `/${slug.join("/")}/` : slug
  const variables = { url }

  log.info(variables, "getStaticProps: invoking at {url}")
  try {
    const client = await getClient("getStaticProps")
    const { data, errors } = await client.query<PagePublicQuery>({
      query: PagePublicDocument,
      variables
    })

    if (errors) {
      throw { graphQLErrors: errors, message: "getStaticProps: GraphQLErrors at {url}" }
    }

    const { page } = data

    if (!page) {
      throw { message: "getStaticProps: no page data at {url}" }
    }

    log.info({ page: page.name, ...variables }, "getStaticProps: finished at {url}")

    return ({ props: { page } })
  }
  catch (error) {
    error.ctx = ctx
    log.error({ ...error, ...variables }, error.message || "getStaticProps: failed at {url}")
    throw error
  }
}
```

All this is for a page route `[...slug].tsx` that lives at the root of the `pages` folder. It causes enough issues in itself, that the `getStatitProps` is called for any files trying to be accessed that are not defined in public or in pages, (my logging service shows errors of not being able to find data in my GraphQL endpoint for `robots.txt` and such.) I guess the solution would be to somehow have at least one folder for the catch-all dynamic route.?

But I think I have an unrelated issue as well. The above implementation results in a bunch of 

```
GET http://localhost:3000/_next/data/n35yCldQmjfUS2fL7ATX6/path/to/article/article.json net::ERR_ABORTED 500 (Internal Server Error)
```
errors in prod console, when data was not found for an article. I thought throwing the error further would be handled by Next.js. Does that not mean that no errors should be shown in this case?Removing the `try catch` block and simply returning `{props: {}}` makes the console error go away, but it is not the wished behavior, as subsequent requests will not try to refetch the needed data.

And even when I go to an article whose data looks to be successfully fetched, I am getting:

```
GET http://localhost:3000/_next/static/n35yCldQmjfUS2fL7ATX6/path/to/article/article.js net::ERR_ABORTED 404 (Not Found)
```
(Note it ends with .js, not .json). Looking at the `_next/static/n35yCldQmjfUS2fL7ATX6/` folder, indeed there is no corresponding .js file. (It is being prefetched by a `link` tag, that I cannot find anywhere in `_next/static`)

What am I doing wrong? If the above code is not enough for answering, I can try creating a reproduction repo.

**Comment by fabarea** [comment]
I am also interested by this topic and asking the same question. How to properly handle errors in getStaticProps / getStaticPaths? 

I would like to throw a 404 page for instance

**Comment by lotharvm** [comment]
@fabarea, this has been mentioned by @timneutkens as a follow-up RFC to the previously implemented 'redirects from GSP'. But yeah totally agree, error handling feels like the last missing piece of SSG/GSP. 

@timneutkens, do you have any update about this?

**Comment by balazsorban44** [comment, marked as answer]
I think this has been sufficiently answered here https://github.com/vercel/next.js/discussions/11180#discussioncomment-2399

Please also look at the discussion below the comment

**Reply by lotharvm** [reply]
To be honest, I'm not sure if this is sufficiently answered in the issue you're referring to. There's still no way to set a 404 statuscode in GSP if, for example, an API call fails. Setting the correct statuscode feels like quite an essential part of error handling.

**Reply by timneutkens** [reply]
Keep an eye on #17755
